omp/base/kernel_launch.hpp

```diff
@@ -36,108 +36,101 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #endif
 
 
+#include "core/synthesizer/implementation_selection.hpp"
+
+
 namespace gko {
 namespace kernels {
 namespace omp {
 
 
-template <typename KernelFunction, typename... KernelArgs>
-void run_kernel(std::shared_ptr<const OmpExecutor> exec, KernelFunction fn,
-                size_type size, KernelArgs&&... args)
-{
-#pragma omp parallel for
-    for (size_type i = 0; i < size; i++) {
-        [&]() { fn(i, map_to_device(args)...); }();
-    }
-}
+namespace {
 
-template <size_type cols, typename KernelFunction, typename... MappedKernelArgs>
-void run_kernel_fixed_cols_impl(std::shared_ptr<const OmpExecutor> exec,
-                                KernelFunction fn, dim<2> size,
-                                MappedKernelArgs... args)
+
+template <typename KernelFunction, typename... MappedKernelArgs>
+void run_kernel_impl(std::shared_ptr<const OmpExecutor> exec, KernelFunction fn,
+                     size_type size, MappedKernelArgs... args)
 {
-    const auto rows = size[0];
 #pragma omp parallel for
-    for (size_type row = 0; row < rows; row++) {
-#pragma unroll
-        for (size_type col = 0; col < cols; col++) {
-            [&]() { fn(row, col, args...); }();
-        }
+    for (int64 i = 0; i < static_cast<int64>(size); i++) {
+        [&]() { fn(i, args...); }();
     }
 }
 
-template <size_type remainder_cols, size_type block_size,
-          typename KernelFunction, typename... MappedKernelArgs>
-void run_kernel_blocked_cols_impl(std::shared_ptr<const OmpExecutor> exec,
-                                  KernelFunction fn, dim<2> size,
-                                  MappedKernelArgs... args)
+
+template <int block_size, int remainder_cols, typename KernelFunction,
+          typename... MappedKernelArgs>
+void run_kernel_sized_impl(syn::value_list<int, remainder_cols>,
+                           std::shared_ptr<const OmpExecutor> exec,
+                           KernelFunction fn, dim<2> size,
+                           MappedKernelArgs... args)
 {
+    const auto rows = static_cast<int64>(size[0]);
+    const auto cols = static_cast<int64>(size[1]);
     static_assert(remainder_cols < block_size, "remainder too large");
-    const auto rows = size[0];
-    const auto cols = size[1];
     const auto rounded_cols = cols / block_size * block_size;
     GKO_ASSERT(rounded_cols + remainder_cols == cols);
+    if (rounded_cols == 0 || cols == block_size) {
+        // we group all sizes <= block_size here and unroll explicitly
```

Comment:
Yes, the remainder part is just necessary for distinguishing between size == block_size and size < block_size, since remainder_cols only tells us the correct unroll size for size < block_size.